Demo: Use grid in dialog as multi-select

Ben-Ho opened the pull request:
For more information see alternative pull-request: #2428 
<details>
    <summary>Screenshots/screencasts</summary>
    
![image](https://github.com/vivid-planet/comet/assets/1324250/470de797-c84b-4a9a-9795-21b9150a91ec)

</details>

nsams commented on demo/admin/src/products/categories/ProductCategoriesPage.tsx:
```diff
@@ -45,7 +65,65 @@ const ProductCategoriesPage: React.FC = () => {
+                                                    },
+                                                }}
+                                            >
+                                                <FinalForm<{ productIds: string[] }>
```
you should not need a form in this dialog

Ben-Ho replied:
alright. then I'll remove this.
But I'm still missing why it's not needed. I thought there needs to be an FinalForm-Component for fields to work (checked DialogContent, it's directly from mui, so no FinalForm... I only found SaveBoundary).
Also how do I perform submit of the mutation?

nsams replied:
> Also how do I perform submit of the mutation?

Should be possible like that:
```
<EditDialog>
    <Savable doSave={...} hasChanges={...} doReset={...} />
    <Grid...>
</EditDialog

```

Ben-Ho replied:
I moved it into a separate form-component because I think the page should not do any mutations but combine different forms, tabs, grids etc.

nsams replied:
Moving it around doesn't make it go away :)

I don't see a reason why you should have a form here, this makes everything much more complicated.

Ben-Ho replied:
I tried it now. There are some issues. Do you also mean to change implementation without a Field? because now it throws an error "useForm must be used inside of a <Form> component"

nsams replied:
Yes, no form and no field. No final-form code anywhere.

```
<EditDialog>
    <Savable doSave={...} hasChanges={...} doReset={...} />
    <Grid...>
</EditDialog
```

Ben-Ho replied:
it's done. no form no field. No final-form code anywhere. :-)

nsams replied:
and, do you think it is better now?

Ben-Ho replied:
Yes definitively. I didn't thought about this option. All implementations in our projects used the final-form approach. So that's probably not the way all developers would solve this, but I like it. And if for some reason it's needed in a form (e.g. mutation also submitting relation-data) it's still possible with a simple/small SelectionGridFieldWrapper.

Ben-Ho commented on demo/admin/src/products/categories/ProductCategoriesPage.tsx:
```diff
@@ -45,7 +65,65 @@ const ProductCategoriesPage: React.FC = () => {
+                                                    mode="edit"
+                                                    onSubmit={async (values, form, event) => {
+                                                        await client.mutate<GQLSetProductCategoryMutation, GQLSetProductCategoryMutationVariables>({
+                                                            mutation: setProductCategoryMutation,
```
I don't think it's best-practice to do a mutation in a page-component, I move this into a separate form-component

nsams commented on demo/api/schema.gql:
```diff
@@ -1166,6 +1166,7 @@ type Mutation {
   createProductCategory(input: ProductCategoryInput!): ProductCategory!
   updateProductCategory(id: ID!, input: ProductCategoryUpdateInput!): ProductCategory!
   deleteProductCategory(id: ID!): Boolean!
+  setProductCategory(id: ID!, input: SetProductCategoryInput!): ProductCategory!
```
instead use the existing updateProductCategory; input is disabled for products relation in ProductCategory entity, just enable it

johnnyomair commented on demo/api/src/products/dto/set-product-category.input.ts:
This file isn't used, please remove.

johnnyomair commented on demo/admin/src/products/categories/ProductsSelectGrid.tsx:
```diff
@@ -0,0 +1,272 @@
+`;
+
+const productsQuery = gql`
+    query ProductsList($offset: Int!, $limit: Int!, $sort: [ProductSort!], $filter: ProductFilter, $search: String) {
```
GraphQL operation names need to be unique across the project, and this names are already used in the generated products grid. This can lead to problems with caching and/or generation once the queries differ (e.g., you remove a field here). Please rename the operations in this file.

nsams replied:
> This can lead to problems with caching

I don't think the query name is used in apollo cache. After all it's a normalisation across all queries, with cache-key type+id

> generation once the queries differ

since we switched to near-operations-file in gql-codegen those queries can have the same name if they are used in different files. The types are generated per file, so there are no conflicts in naming.

----

The only reason for unique query names might be for profiling or debugging on the server side (so you get a hint where the query comes from)

(NOTE: fragments have to be unique, I don't know why though)

johnnyomair commented on demo/admin/src/products/categories/ProductsSelectGrid.tsx:
```diff
@@ -0,0 +1,272 @@
+}
+
+const productsFragment = gql`
+    fragment ProductsListManual on Product {
```
Please remove the fields here that aren't used in the grid.

johnnyomair commented on demo/admin/src/products/categories/AssignProductsForm.tsx:
```diff
@@ -0,0 +1,79 @@
+            initialValuesEqual={isEqual} //required to compare block data correctly
+            subscription={{}}
+        >
+            {({ handleSubmit, submitting }) => {
```
Neither of these are used, could we remove them?

johnnyomair commented on demo/admin/src/products/categories/AssignProductsForm.tsx:
```diff
@@ -0,0 +1,79 @@
+        >
+            {({ handleSubmit, submitting }) => {
+                return (
+                    <Field name="productIds" fullWidth>
+                        {(props) => (
+                            // TODO: height should be defined by parent, but is required because of nesting in field. use FinalFormField directly?
+                            <Box sx={{ height: "70vh" }}>
+                                <ProductsSelectGrid {...props} />
+                            </Box>
+                        )}
+                    </Field>
```
```suggestion
                    <Field name="productIds" component={ProductsSelectGrid} fullWidth />
```

Why isn't this possible?

Ben-Ho replied:
styling issues. the grid does not have a height and the height of the form is not forwarded into the fields because how FieldContainer is styled/implemented.

![image](https://github.com/vivid-planet/comet/assets/1324250/1569f7e9-e14f-4128-8c24-9b2b290ef280)

johnnyomair commented on demo/admin/src/products/categories/ProductCategoriesPage.tsx:
```diff
@@ -45,7 +52,40 @@ const ProductCategoriesPage: React.FC = () => {
+                                        <Box sx={{ height: "100vh" }}>
+                                            <ProductsGrid filter={{ category: { equal: selectedId } }} />
+                                            {/* TODO change button behaviour and open edit-dialog, open discussion https://github.com/vivid-planet/comet/pull/2171 */}
+                                            <EditDialog
```
Shouldn't the EditDialog be defined inside AssignProductsForm?

Ben-Ho replied:
In my opinion no, why should a form include it's dialog? a form should be usable everywhere. a form should not know or define where it's rendered. Maybe I could add an additional component "AssignProductsDialogForm" combining EditDialog and AssignProductsForm. But I don't think moving this 5 lines of code would bring any benefit.

nsams commented on demo/admin/src/products/categories/AssignProductsGrid.tsx:
```diff
@@ -0,0 +1,78 @@
+export function AssignProductsGrid({ productCategoryId }: FormProps): React.ReactElement {
+    const client = useApolloClient();
+
+    const [selection, setSelection] = useState<{ initialValues?: string[]; values: string[] }>({ values: [] });
```
please split that into two states. And for initialValues I would use useQuery

Ben-Ho replied:
I tried this in my first attempt, but I had no idea how to use data from useQuery as initial-value for values-state in a clean way. what do you think about this:
```
    const { data, error, loading } = useQuery<GQLGetProductIdsForProductCategoryQuery, GQLGetProductIdsForProductCategoryQueryVariables>(
        getProductIdsForProductCategory,
        {
            variables: { id: productCategoryId },
        },
    );
    const initialValues = data?.products.nodes ? data.products.nodes.map((product) => product.id).sort() : [];
    const [values, setValues] = useState<string[]>(initialValues);
    useEffect(() => {
        setValues(initialValues);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [data?.products.nodes]);

    if (error) return <FormattedMessage id="common.error" defaultMessage="An error has occured. Please try again at later" />;
    if (loading) return <CircularProgress />;
```

okay, now that I see it I think it's ok. Do you know any trick to avoid the disable-react-hooks/exhaustive-deps? using initialValues directly does obviously always trigger a state-change...

johnnyomair commented on demo/admin/src/products/categories/AssignProductsGrid.tsx:
```diff
@@ -0,0 +1,80 @@
+            variables: { id: productCategoryId },
+        },
+    );
+    const initialValues = data?.products.nodes ? data.products.nodes.map((product) => product.id).sort() : [];
```
Why do you need to sort the array?

Ben-Ho replied:
this is required for `hasChanges={!isEqual(...)}` to work correctly. else deselecting/selecting the same item is detected as change and closing the dialog does ask to save changes even if it hasn't really changed.

johnnyomair replied:
IMO it would be better to only sort in the hasChanges prop, which would make it more obvious why you sort. Or you could write a custom array comparison and don't use a deep equality check here

johnnyomair commented on demo/admin/src/products/categories/AssignProductsGrid.tsx:
```diff
@@ -0,0 +1,80 @@
+    );
+    const initialValues = data?.products.nodes ? data.products.nodes.map((product) => product.id).sort() : [];
+    const [values, setValues] = useState<string[]>(initialValues);
+    useEffect(() => {
+        setValues(initialValues);
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [data?.products.nodes]);
```
Why do you need this effect? Disabling the ESLint rule doesn't seem like a good idea to me.

Ben-Ho replied:
because fetched values need to be in values-state to allow deselecting. `data.products.node` is not set for first render so the default-value for values-useState is an empty array and is not updated after `data.products.node` is finally fetched. es-lint requires useEffect to contain initialValues in it's deps-array, but this is always newly initialized resulting in endless re-render.

johnnyomair replied:
It would be better to do this this way:

```ts
const initialValues = useMemo(() => {
  return data?.products.nodes ? data.products.nodes.map((product) => product.id).sort() : [];
}, [data]);

useEffect(() => {
  setValues(initialValues)
}, [initialValues]);
```

Or, remove the useEffect altogether:

```ts
const initialValues = data?.products.nodes ? data.products.nodes.map((product) => product.id).sort() : [];

if (!isEqual(initialValues, values)) {
  setValues(initialValues);
}
```

BTW, this is a conservation I would like to resolve 😉

Ben-Ho replied:
tried your second idea but it didn't work because !isEqual resulted in always reseting back to initialValues, so used your first suggestion.

johnnyomair reviewed (commented):
We can't merge this PR before we fix the dialog behavior

Ben-Ho commented:
> We can't merge this PR before we fix the dialog behavior

I switched to a custom dialog, but it's still missing the "close-dialog-unsaved-changes-check".

Ben-Ho commented:
> We can't merge this PR before we fix the dialog behavior

@johnnyomair what do you mean by that? Implementing the mentioned SaveDialog in comet or just switching to a handmade custom dialog with this pull-request, missing some features?

johnnyomair commented:
> what do you mean by that?

This was before you switched to a custom dialog, where the dialog opened upon navigating to the tab. Should be fine now.

johnnyomair commented:
I just tried this locally and found some UX issues:

1. The grid doesn't refresh after adding new products
2. Removing a product from the grid deletes the product. I'd expect the assignment to be removed
3. No discard changes dialog (you already mentioned that)

IMO 1 and 2 should be addressed before merging.

Ben-Ho commented:
@johnnyomair @nsams one aspect of this pull-request is providing a demo how the grid should be implemented. I think this part is ok? So I could start add this feature to our admin-generator?

should this be a completely different type (e.g. "form", "grid", "selectGrid") or a variance of "grid". (e.g. type: {... "grid", usage: "selectGrid", ...}). I think a different type, because many grid-features don't make sense, but implemented like a variance, using generate-grid-code with fixed/predefined/simplified grid-config because it's the same in many ways...

nsams commented on demo/admin/src/products/categories/ProductCategoriesPage.tsx:
```diff
@@ -45,7 +56,69 @@ const ProductCategoriesPage: React.FC = () => {
+                                                }
+                                                filter={{ category: { equal: selectedId } }}
+                                            />
+                                            <SaveBoundary onAfterSave={() => setIsOpen(false)}>
```
Please move this (ProductsGrid+SaveBoundary+Dialog etc) into it's own Component (and file)

ProductCateogriesPage should only combine together components and not have any logic on it's own.

Ben-Ho replied:
I'd rather only move SaveBoundary+Dialog into a SaveDialog-component (EditDialog does not work in this case). Because I would expect ProductCateogriesPage to decide how the structure+navigation is by combining ProductsGrid + AssignProductsGrid (by wrapping in SaveDialog) in ProductCateogriesPage. (maybe also wrap ProductsGrid in AssignedProductsGrid-component, to make it more obvious and hide the filtering away. If it wasn't this weird ManyToOne-situation there would probably be already a generated AssignedProductsGrid component)

additionally the page-component could be further split if it becomes to nested... (e.g. ProductCategoryDetailPage, ProductCategoryAssignedProductsRouterTab)

nsams commented:
> should this be a completely different type (e.g. "form", "grid", "selectGrid") or a variance of "grid". (e.g. type: {... "grid", usage: "selectGrid", ...}).

please give two examples of how the config could look like

johnnyomair commented:
> The grid doesn't refresh after adding new products

This could probably be fixed by using the refetchQueries option of the mutation.

johnnyomair commented:
>  one aspect of this pull-request is providing a demo how the grid should be implemented. I think this part is ok? So I could start add this feature to our admin-generator?

If it's part of the handmade demo you could address my second UX concern first

Ben-Ho commented on demo/admin/src/products/categories/ProductCategoriesPage.tsx:
```diff
@@ -45,7 +56,69 @@ const ProductCategoriesPage: React.FC = () => {
+                                                        <AssignProductsGrid productCategoryId={selectedId} />
+                                                    </DialogContent>
+                                                    <DialogActions>
+                                                        {/* TODO Missing close-dialog-unsaved-changes-check */}
```
EditDialog had some issues in this situation. @thomasdax98 told me to use something else because EditDialog was not meant to be used for something like that.

Ben-Ho commented:
> > should this be a completely different type (e.g. "form", "grid", "selectGrid") or a variance of "grid". (e.g. type: {... "grid", usage: "selectGrid", ...}).
> 
> please give two examples of how the config could look like

```
const config: GridConfig<ProductCategory> = {
    type: "selectGrid",
    fragmentName: "ProductCategory",
    query: "productCategories", // generator should throw an exception if query supports paging
    columns: [...],
    filterProp: true, // should be usable
    rowActionProp: false, // maybe usefull, e.g. to open dialog to check row-details
}
```
vs
```
const config: GridConfig<ProductCategory> = {
    type: "grid",
    usage: "selectGrid",
    fragmentName: "ProductCategory",
    query: "productCategories", // generator should throw an exception if query supports paging
    columns: [...],
    readOnly: true, // should not be able to set to false, or will have no effect
    
    add: false, // will have no effect, due to readOnly: true
    edit: false, // will have no effect, due to readOnly: true
    delete: false, // will have no effect, due to readOnly: true
    copyPaste: false, // will have no effect, due to readOnly: true
    
    toolbar: false, // should not be able to set to true, or will have no effect
    toolbarActionProp: false, // should not be able to set to true, or will have no effect

    filterProp: true, // should be usable
    rowActionProp: false, // maybe usefull, e.g. to open dialog to check row-details
}
```

Ben-Ho commented:
> > The grid doesn't refresh after adding new products
> 
> This could probably be fixed by using the refetchQueries option of the mutation.

the used grid is generated, so this requires changes to admin-generator. should I open a pull-request adding the option to forward refetchQueries option?

Ben-Ho commented:
> > one aspect of this pull-request is providing a demo how the grid should be implemented. I think this part is ok? So I could start add this feature to our admin-generator?
> 
> If it's part of the handmade demo you could address my second UX concern first

both concerns are related to the generated grid. for the second concern I got this options:

- should it be possible to switch delete-button behavior, or provide custom actions via props for the generated grid
- Or build a fake-resolver to generate a grid especially for this case?

Ben-Ho commented:
> > > The grid doesn't refresh after adding new products
> > 
> > 
> > This could probably be fixed by using the refetchQueries option of the mutation.
> 
> the used grid is generated, so this requires changes to admin-generator. should I open a pull-request adding the option to forward refetchQueries option?

ah, now I know what you mean by refetchQueries. That's really convenient. :-D maybe not the cleanest in terms of predictability, because it may be hard to find the source for refetching, but if used sparsely it's much easier than any other solution.

Ben-Ho commented:
@johnnyomair I'm sorry, I thought I'd be using the generated products-grid but it's actually the handmade. fixed both concerns. 

- The grid doesn't refresh after adding new products 
   - solved with apollo-client refetch-queries feature, needed to export the productsQuery. I thought this would be cleaner than by query-name.
- Removing a product from the grid deletes the product. I'd expect the assignment to be removed
   - I added a prop to optionally exclude CrudContextMenu-options.

johnnyomair commented on demo/admin/src/products/ProductsGrid.tsx:
```diff
@@ -54,20 +54,20 @@ function ProductsGridToolbar() {
+    toolbarAction?: React.ReactNode;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    rowAction?: (params: GridRenderCellParams<any, GQLProductsListManualFragment, any>) => React.ReactNode;
+    hideCrudContextMenu?: boolean;
```
Yes, that's what I meant. But we should discuss if this is a good idea first 😅

johnnyomair replied:
Shouldn't this be readonly instead? As in the generator?

Ben-Ho replied:
this is now solved in #2428 by creating a copy of productsGrid and modifying. I'll postpone the change for this pull-request because I assume we will go with #2428

johnnyomair commented:
As discussed, please implement an alternative version that uses the row selection model: https://mui.com/x/react-data-grid/row-selection/#controlled-row-selection

Ben-Ho commented:
> As discussed, please implement an alternative version that uses the row selection model: https://mui.com/x/react-data-grid/row-selection/#controlled-row-selection

here's the demo: #2428
and here the admin-generator adjustment: #2449

Ben-Ho commented:
I reminded myself we said this can be closed in favor of #2449 because admin-generator generates best-practice (=demo)
